### PR TITLE
test/librados: free AioCompletion using AioCompletion::release()

### PR DESCRIPTION
**Pull request by tchaikov**
this is the supported way for destroying an `AioCompletion` if the
`AioCompletion` is created using `Rados::aio_create_completion()`.

if we simply `delete` a pointer of `AioCompletion`, the
`AioCompletion::pc` would be leaked.

the tests also serve as samples for potential developers of librados
applications, so it's important to make sure it shows how the APIs are
supposed to be used.

Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by liewegas** [comment]
I wonder if we should (have) just made ``~AioCompletion`` do what release does.  At this point, we could just make ``void AioCompletion::release() { delete this; }`` and move the old ``release()`` body to ``~AioCompletion``...?  (Right now, it's ```
void librados::AioCompletion::AioCompletion::release()
{
  AioCompletionImpl *c = (AioCompletionImpl *)pc;
  c->release();
  delete this;
}
```

**Comment by tchaikov** [comment]
@liewegas updated and repushed.

**Comment by tchaikov** [comment]
> although I'd recast this as 2 patches, one that fixes the dtor, and the second that updates the tests to unique_ptr

@liewegas makes sense to me! rebased and repushed.

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
retest this please

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
retest this please.

**Comment by liewegas** [comment]
```
2019-09-09T15:36:45.305 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio: [ RUN      ] LibRadosAio.AioUnlock
2019-09-09T15:36:45.306 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/15.0.0-4717-gb368246/rpm/el7/BUILD/ceph-15.0.0-4717-gb368246/src/com
mon/mutex_debug.h: In function 'ceph::mutex_debug_detail::mutex_debug_impl<<anonymous> >::~mutex_debug_impl() [with bool Recursive = false]' thread 7f0c0b23c700 time 2019-09-09T15:36:45.304189+0000
2019-09-09T15:36:45.313 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/15.0.0-4717-gb368246/rpm/el7/BUILD/ceph-15.0.0-4717-gb368246/src/com
mon/mutex_debug.h: 111: FAILED ceph_assert(r == 0)
2019-09-09T15:36:45.313 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  ceph version 15.0.0-4717-gb368246 (b3682463a7b7edfb2df8e0780cfcc187a6e4e1ad) octopus (dev)
2019-09-09T15:36:45.314 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14a) [0x7f0c12a07e8f]
2019-09-09T15:36:45.314 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  2: (()+0x26c057) [0x7f0c12a08057]
2019-09-09T15:36:45.314 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  3: (()+0x5beb6) [0x7f0c1be84eb6]
2019-09-09T15:36:45.314 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  4: (()+0x93db1) [0x7f0c1bebcdb1]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  5: (()+0x59f19) [0x7f0c1be82f19]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  6: (()+0xc9344) [0x7f0c1bef2344]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  7: (()+0xc9ccb) [0x7f0c1bef2ccb]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  8: (()+0xe9922) [0x7f0c1bf12922]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  9: (DispatchQueue::fast_dispatch(boost::intrusive_ptr<Message> const&)+0x6aa) [0x7f0c12c1aa6a]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  10: (ProtocolV2::handle_message()+0x12e1) [0x7f0c12ce2aa1]
2019-09-09T15:36:45.315 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  11: (ProtocolV2::handle_read_frame_dispatch()+0x220) [0x7f0c12cf2820]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  12: (ProtocolV2::handle_read_frame_epilogue_main(std::unique_ptr<ceph::buffer::v14_2_0::ptr_node, ceph::buffer::v14_2_0::ptr_node::disposer>&&, int)+0x44d) [0x7f0c12cf2c8d]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  13: (ProtocolV2::run_continuation(Ct<ProtocolV2>&)+0x34) [0x7f0c12cdbf34]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  14: (AsyncConnection::process()+0x794) [0x7f0c12ca99e4]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  15: (EventCenter::process_events(unsigned int, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >*)+0xa25) [0x7f0c12cf9d15]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  16: (()+0x565927) [0x7f0c12d01927]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  17: (()+0x83412f) [0x7f0c12fd012f]
2019-09-09T15:36:45.316 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  18: (()+0x7dd5) [0x7f0c1bc14dd5]
2019-09-09T15:36:45.317 INFO:tasks.workunit.client.0.smithi136.stdout:                  api_aio:  19: (clone()+0x6d) [0x7f0c1099202d]
2019-09-09T15:36:45.317 INFO:tasks.workunit.client.0.smithi136.stderr:bash: line 1: 15301 Aborted                 (core dumped) ceph_test_rados_api_aio 2>&1
```
/a/sage-2019-09-09_13:13:58-rados-wip-sage-testing-2019-09-08-2045-distro-basic-smithi/4293114

File: src/librados/librados_c.cc [hunk 1]
```diff
@@ -3024,6 +3024,7 @@ extern "C" int _rados_aio_unlock(rados_ioctx_t io, const char *o, const char *na
   librados::IoCtx ctx;
   librados::IoCtx::from_rados_ioctx_t(io, ctx);
   librados::AioCompletionImpl *comp = (librados::AioCompletionImpl*)completion;
+  comp->get();
```
Comment (tchaikov) [inline src/librados/librados_c.cc]: we decrease the refcount of inner impl pointer in destructor of `AioCompletion`, so we should hold a strong reference .

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-19_09:56:38-rados-wip-kefu2-testing-2019-11-19-1509-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-21_08:41:20-rados-wip-kefu2-testing-2019-11-19-1509-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-23_01:15:32-rados-wip-kefu2-testing-2019-11-23-0140-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-24_13:33:51-rados-wip-kefu2-testing-2019-11-24-1912-distro-basic-mira/

**Pull request merged**